pkg/ipcache/ipcache.go

```diff
@@ -0,0 +1,213 @@
+// Copyright 2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ipcache
+
+import (
+	"encoding/json"
+	"path"
+	"sort"
+	"sync"
+
+	"github.com/cilium/cilium/pkg/envoy"
+	envoyAPI "github.com/cilium/cilium/pkg/envoy/cilium"
+	"github.com/cilium/cilium/pkg/envoy/xds"
+	"github.com/cilium/cilium/pkg/identity"
+	"github.com/cilium/cilium/pkg/kvstore"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	// DefaultAddressSpace is the address space used if none is provided.
+	DefaultAddressSpace = "default"
+)
+
+var (
+	// IPIdentitiesPath is the path to where endpoint IPs are stored in the key-value
+	//store.
+	IPIdentitiesPath = path.Join(kvstore.BaseKeyPrefix, "state", "endpointIPs", "v1")
+
+	// IPIdentityCache caches the mapping of endpoint IPs to their corresponding
+	// security identities across the entire cluster in which this instance of
+	// Cilium is running.
+	IPIdentityCache = NewIPCache(envoy.NetworkPolicyHostsCache)
+
+	// AddressSpace is the address space (cluster, etc.) in which policy is
+	// computed. It is determined by the orchestration system / runtime.
+	AddressSpace = DefaultAddressSpace
+
+	setupIPIdentityWatcher sync.Once
+)
+
+// IPCache is a caching of endpoint IP to security identity (and vice-versa) for
+// all endpoints which are part of the same cluster.
+type IPCache struct {
+	mutex              lock.RWMutex
+	ipToIdentityCache  map[string]identity.NumericIdentity
+	identityToIPCache  map[identity.NumericIdentity]map[string]struct{}
+	xdsResourceMutator xds.ResourceMutator
+}
+
+// NewIPCache returns a new IPCache with the mappings of endpoint IP to security
+// identity (and vice-versa) initialized, along with the provided resourceMutator.
+func NewIPCache(resourceMutator xds.ResourceMutator) *IPCache {
+	return &IPCache{
+		ipToIdentityCache:  map[string]identity.NumericIdentity{},
+		identityToIPCache:  map[identity.NumericIdentity]map[string]struct{}{},
+		xdsResourceMutator: resourceMutator,
+	}
+}
+
+// Upsert adds / updates  the provided IP and identity into both caches contained
+// within ipc.
+func (ipc *IPCache) Upsert(endpointIP string, identity identity.NumericIdentity) {
```

Comment:
Make this private, in that case, the watcher is guaranteed to be the only writer into the cache and the locking you have is sufficient.